js/services/contact_service.js

```diff
@@ -8,6 +8,13 @@ angular.module('contactsApp')
 	var observerCallbacks = [];
 	var loadPromise = undefined;
 
+	var allUpdates = $q.when();
+	this.AddAndUpdate = function(contact) {
+		allUpdates = allUpdates.then(function() {
+			return contactService.update(contact);
+		});
+	};
+
```

Comment:
Nice technique to serialize updates!

A few remarks:
* we could think about using the browser in-built Promises instead of Angular's `$q`. It is supported by all major browsers for some time already.
* why did you call the function `AddAndUpdate`? All the other functions are called `update`. Also it should be lowercase 😜 
* could we make this part of the update method? This would also prevent any misuse in the future.

Reply:
I found this technique on a stack somewhere! I liked it a lot! Well the addAnsUpdate is here as a secondary function. We can't really use the same update function name? :)

Reply:
Agreed about the browser promises, I mostly went to the angular way since it was already used here. ;)

Reply:
> We can't really use the same update function name?

but why don't you apply this technique directly inside the update function, instead of in a separate wrapper function?

Reply:
That's a really good question 😂

Reply:
Hum, I have an answer for that finally, It's easier to understand! :)

Reply:
> we could think about using the browser in-built Promises instead of Angular's $q. It is supported by all major browsers for some time already.

$q had the advantage to be linked to the $digest process, so we shouldn't get rid of it :)
https://stackoverflow.com/a/34516118/3885878